### ISSUE TITLE
brew error on M1 with macOS Monterey 12.3

### ISSUE THREAD

**Issue by Leeiio** [TomAnthony/itermocil#121]
I'm getting this.

```
→ brew install TomAnthony/brews/itermocil
==> Downloading https://pypi.python.org/packages/source/P/PyYAML/PyYAML-3.11.tar.gz
Already downloaded: /Users/Leeiio/Library/Caches/Homebrew/downloads/d98de4b2d4d986c9eacf67d572ad3182da5555a27ae39f38260d899047b9ef4b--PyYAML-3.11.tar.gz
==> Downloading https://github.com/TomAnthony/itermocil/archive/0.2.1.tar.gz
Already downloaded: /Users/Leeiio/Library/Caches/Homebrew/downloads/7660c48dfb6cb50ac68fa581c5de591c2db93fb698776e83084c669a45eb0299--itermocil-0.2.1.tar.gz
==> Installing itermocil from tomanthony/brews
==> python -c import setuptools... --no-user-cfg install --prefix=/opt/homebrew/Cellar/itermocil/0.2.1/libexec/vendor --install-scripts=/opt/homebrew/Cellar/itermocil/0.2.1/libexec/vendor/bin --single-version-externally-managed --r
Last 15 lines from /Users/Leeiio/Library/Logs/Homebrew/itermocil/01.python:
2022-03-16 04:59:53 +0000

python
-c
import setuptools, tokenize
__file__ = 'setup.py'
exec(compile(getattr(tokenize, 'open', open)(__file__).read()
  .replace('\r\n', '\n'), __file__, 'exec'))
--no-user-cfg
install
--prefix=/opt/homebrew/Cellar/itermocil/0.2.1/libexec/vendor
--install-scripts=/opt/homebrew/Cellar/itermocil/0.2.1/libexec/vendor/bin
--single-version-externally-managed
--record=installed.txt


If reporting this issue please do so at (not Homebrew/brew or Homebrew/core):
  https://github.com/tomanthony/homebrew-brews/issues

These open issues may also help:
I can't install itermocil.  https://github.com/TomAnthony/homebrew-brews/issues/4
itermocil install is failing https://github.com/TomAnthony/homebrew-brews/issues/2
```

**Comment by scottluxford** [comment]
[Python support was dropped from MacOS 12.3](https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python). The script is pointing at /usr/bin/python which is now kaput!

I'll update if I find a decent solution.

**Comment by zvictor** [comment]
Is there a temporary workaround?

**Comment by AdrienPoupa** [comment]
I am using the following very hacky, very dirty workaround:

1. Install Python 2.7.18 through pyenv, given Homebrew no longer supports it:
```
> brew install pyenv
> pyenv install 2.7.18
```
2. Replace the Python2 bin location in itermocil binary:
```
> nano /usr/local/Cellar/itermocil/0.2.1/libexec/bin/itermocil
```
```diff
-#!/usr/bin/python
+#!/Users/YOUR_USERNAME/.pyenv/versions/2.7.18/bin/python2.7
```

Now the proper solution would be to migrate the codebase to Python3 and change the Python binary location, but I don't see that happening anytime soon unfortunately.

**Comment by Leeiio** [comment]
I have uninstall the itermocil and now I can not install it any more...

**Comment by godbout** [comment]
just in case. as this repo doesn't seem to be updated anymore, i've forked and updated for Python3: https://github.com/godbout/itermocil

**Comment by scottluxford** [comment]
Nice one @AdrienPoupa - this was my approach too.

But @godbout - legend. Your fork works a treat. Cheers.

**Comment by godbout** [comment]
@scottluxford enjoy 💫  had to get this fixed before working on my projects. life without iTermocil is so sparkless.

**Comment by Leeiio** [comment]
> just in case. as this repo doesn't seem to be updated anymore, i've forked and updated for Python3: https://github.com/godbout/itermocil

Nice job! But when I install the itermocil and use `itermocil test`, I am getting this

```
Traceback (most recent call last):
  File "/Users/Leeiio/.pyenv/versions/3.7.12/bin/itermocil", line 11, in <module>
    load_entry_point('itermocil==0.2.3', 'console_scripts', 'itermocil')()
  File "/Users/Leeiio/.pyenv/versions/3.7.12/lib/python3.7/site-packages/itermocil.py", line 752, in main
    instance = Itermocil(filepath, here=args.here, cwd=cwd)
  File "/Users/Leeiio/.pyenv/versions/3.7.12/lib/python3.7/site-packages/itermocil.py", line 79, in __init__
    self.process_file()
  File "/Users/Leeiio/.pyenv/versions/3.7.12/lib/python3.7/site-packages/itermocil.py", line 568, in process_file
    self.arrange_panes(len(window['panes']), layout)
  File "/Users/Leeiio/.pyenv/versions/3.7.12/lib/python3.7/site-packages/itermocil.py", line 262, in arrange_panes
    for p in range(0, second_columns):
TypeError: 'float' object cannot be interpreted as an integer
```

Anyone know how to solve this? I have zero knowledge in 🐍

**Comment by godbout** [comment]
no idea sorry. zero Python here also. was not even aware that there were tests?
in my case i'm under Python 3.9. it's possible that the changes i've made are not enough for your config. they work for my owns with setting up vertical and horizontal panes though, and multiple tabs.

**Comment by godbout** [comment]
you have the stack in the error log, and the error shows an issue between float and integer conversion. that should help. maybe an issue with your config, maybe some more changes to be made. strings needed utf8 encoding/decoding. possible that there's more type restrictions now.

**Comment by Leeiio** [comment]
open itermocil.py file, change the code in line 255 `second_columns = num_panes / 2` to `second_columns = num_panes // 2`, done.

**Comment by godbout** [comment]
nice :D do a PR!

**Comment by Lzmog** [comment]
> open itermocil.py file, change the code in line 236 `second_columns = num_panes / 2` to `second_columns = num_panes // 2`, done.

I would like to point out, that first of all 
- you have to install itermocil from this repo: https://github.com/TomAnthony/itermocil/issues/121#issuecomment-1069266639
- Then change code in line 236
-  **And only then install itermocil**
``` python3 setup.py install --install-scripts=/usr/local/bin ```

**Comment by godbout** [comment]
ok so if no one does it i'll do it.
well i did it.

**Comment by godbout** [comment]
> ok so if no one does it i'll do it.
> well i did it.

huh i saw the PR made by @Leeiio only after i've made the changes :D sorry, didn't receive any notification. strange. anyways, still merged. i've made the changes necessary regarding this issue for the other layouts too.

**Comment by godbout** [comment]
@Leeiio i'll check. maybe if you're using my fork it's best to report bugs there.

**Comment by godbout** [comment]
> windows:
>   - name: test
>     root: ~
>     layout: tiled
>     panes:
>       - commands:
>           - ls
>       - commands:
>           - ls
>       - commands:
>           - ls
>         focus: true
>       - commands:
>           - ls

works here.

<img width="1280" alt="Screen Shot 2022-03-19 at 20 47 15" src="https://user-images.githubusercontent.com/121373/159121702-145f6d9c-7213-4a16-ad27-b749d3667b1b.png">

are you sure you're using the right iTermocil? if you make changes yourself in the `itermocil.py` file you need to run the file with `python3 itermocil.py` unless you run the setup each time.

**Comment by Leeiio** [comment]
@godbout Thanx, I have reinstalled the iTermocil, now everything is ok!

**Comment by DougThwaites** [comment]
@godbout thanks so much for doing this, worked like a charm.  Do you have a buymeacoffee.com link or something?

**Comment by godbout** [comment]
@DougThwaites if you're into Vim i built Vim Mode for macOS. license is one coffee a month: https://github.com/godbout/kindaVim.theapp and https://kindavim.app
if you're not into Vim, then i have an old ko-fi account yeah but i might use the money to buy beers instead: http://ko-fi.com/A1553NO
else thoughts and prayers for GitHub to accept my sponsorship account. i've only been waiting for 2.5 years.

thank you!

**Comment by DougThwaites** [comment]
Not really a Vim guy but looks cool.

Just chucked a few bob your way via Ko-fi, hope it'll get ya a couple of beers where you are.

Also that's crazy about the sponsorship account taking so long. Good luck.

**Comment by godbout** [comment]
@DougThwaites that one's on you:

![IMG_2389](https://user-images.githubusercontent.com/121373/160331136-c7eb17fe-9ef7-4b56-922c-776f9266baa5.jpg)

(hopefully i'm better at building software than i am at taking pictures.)

thanks a lot. greatly appreciated!

**Comment by godbout** [comment]
and of course i'll update this thread with the next beer.

**Comment by Lzmog** [comment]
Found another solution. 
https://github.com/mapbox/node-sqlite3/issues/1552#issuecomment-1075059490

**Comment by ruudk** [comment]
@Lzmog Doesn't this change the default Python to 2.7.18 everywhere? Won't this break other apps?

**Comment by Lzmog** [comment]
**- Doesn't this change the default Python to 2.7.18 everywhere?**
           _As I understand, yes it changes the default Python into 2.7.18._

**- Won't this break other apps?**
           _As it is now, I didn't notice any breaks related to other apps, but it might effect some apps in future._

**Comment by godbout** [comment]
> and of course i'll update this thread with the next beer.

as promised. 

![C8C59CC8-73AD-4DE4-8DB1-16A20D13C267](https://user-images.githubusercontent.com/121373/160779031-c9f3b9e9-348e-4026-bdb2-365a8997f32a.jpeg)

Erdinger during lunch break at a client (Wed afternoon are reserved for clients. all the other half days are for open source). but 10$. i think i overdid it. ☹️

**Comment by godbout** [comment]
> Found another solution. [mapbox/node-sqlite3#1552 (comment)](https://github.com/mapbox/node-sqlite3/issues/1552#issuecomment-1075059490)
> 
> 1. Install pyenv with `brew install pyenv`
> 2. Install Python 2.7 with `pyenv install 2.7.18` and then `pyenv global 2.7.18`
> 3. Then set the default paths typing in the console `export PATH="${HOME}/.pyenv/shims:${PATH}"`

personally i would avoid installing Python 2.8 on a system. it’s not supported anymore and Python 3 came out in two thousand fucking EIGHT (yes eight, not eighteen).

**Comment by DougThwaites** [comment]
Amazing @godbout, thanks again for your work mate.

**Comment by godbout** [comment]
@DougThwaites hey no worries man. i did it selfishly for myself first. and it was just a couple of tweaks. best ration of work over love received! thanks again for the words and the 🍺

**Comment by henryalps** [comment]
Additional point. With @godbout 's solution (thx!) I got this 
`error: The 'PyYAML' distribution was not found and is required by iTermocil`

I tried to remove line 1 `PyYAML==3.11` in file requirements.txt and line 21 `PyYAML` in file setup.py. Now it works like a magic

**Comment by shreyas-s-rao** [comment]
> I tried to remove line 1 PyYAML==3.11 in file requirements.txt and line 21 PyYAML in file setup.py. Now it works like a magic

@henryalps This allowed installation to to go through, but executing the `itermocil` command still throws:
```
Traceback (most recent call last):
  File "/usr/local/bin/itermocil", line 33, in <module>
    sys.exit(load_entry_point('iTermocil==1.0.0', 'console_scripts', 'itermocil')())
  File "/Users/i349079/Downloads/itermocil/itermocil.py", line 733, in main
    instance = Itermocil(filepath, here=args.here, cwd=cwd)
  File "/Users/i349079/Downloads/itermocil/itermocil.py", line 53, in __init__
    self.parsed_config = yaml.load(f, Loader=yaml.Loader)
  File "/Users/i349079/Library/Python/3.8/lib/python/site-packages/yaml/__init__.py", line 79, in load
    loader = Loader(stream)
  File "/Users/i349079/Library/Python/3.8/lib/python/site-packages/yaml/loader.py", line 44, in __init__
    Reader.__init__(self, stream)
  File "/Users/i349079/Library/Python/3.8/lib/python/site-packages/yaml/reader.py", line 85, in __init__
    self.determine_encoding()
  File "/Users/i349079/Library/Python/3.8/lib/python/site-packages/yaml/reader.py", line 124, in determine_encoding
    self.update_raw()
  File "/Users/i349079/Library/Python/3.8/lib/python/site-packages/yaml/reader.py", line 178, in update_raw
    data = self.stream.read(size)
TimeoutError: [Errno 60] Operation timed out
```
The same error is thrown even if I try installing PyYAML and keeping it in the requirements.
Every other method failed for me. I've been so used to this tool on my old Mac for the past 3 years, that it will now become tough for me to use my Mac without it! XD

**Comment by godbout** [comment]
no idea if it's related to your issue but just in case if you're on an Apple silicon the install folder should be: `/opt/homebrew/`

hence if you're installing from my fork: `python3 setup.py install --install-scripts=/opt/homebrew/bin`

i'll update the README.

**Comment by shreyas-s-rao** [comment]
@godbout Thanks a lot for the suggestion. But this did not work too. Installation goes through fine, but the same runtime error still pops up. I tried uninstalling and re-installing PyYAML too, to the needed version (`3.11`), but no luck there as well.

**Comment by ruudk** [comment]
@TomAnthony Would it be possible to give somebody from this thread access to this repository so that we can just fix this one instead of having to do all kinds of hacks/forks to get it working again? Itermocil is such a great piece of software that I used every day and can no longer. 💙

**Comment by godbout** [comment]
> @godbout Thanks a lot for the suggestion. But this did not work too. Installation goes through fine, but the same runtime error still pops up. I tried uninstalling and re-installing PyYAML too, to the needed version (`3.11`), but no luck there as well.

just in case, i think PyAML is needed and you should put it back, if you're removed that line. not sure if it's a Python 3.8 error. i got a fresh install from last week on 12.3 and it worked fine. Python 3.9 (latest) installed from Homebrew.

**Comment by TomAnthony** [comment]
@ruudk - I have added you. I haven't had time to maintain this for sometime now, and have reached out to a few people before to find a new maintainer without much luck. Happy if someone on this thread wants to take a wider maintainer role! :)

**Comment by ruudk** [comment]
@TomAnthony Thanks for your reply! My Python skills are terrible. 

@godbout Since you have been so helpful in this thread, would this be something for you to take on? I would be happy to offer you a few beers of coffee's 😊

**Comment by DougTGP** [comment]
@godbout I'd chuck another few beers in the pot along with @ruudk to get the original repo all ticky boo.

**Comment by godbout** [comment]
@DougThwaites i definitely can help with making the official repo up to date. i forked before instead because the repo didn’t seem maintained anymore. for future PRs tho i don’t think i’ll have the Python skills to judge. PHP JS Swift here and going full Mac apps dev so python is definitively not on the list. if it’s fine with you guys then i’m in.

**Comment by ruudk** [comment]
@godbout That's great! We just need somebody to make it working again on Python 3. 

@TomAnthony Could you invite @godbout ? Thanks!

**Comment by TomAnthony** [comment]
@ruudk @godbout Done! :)

**Comment by godbout** [comment]
guys. i’m in the middle of applying for my French passport renewal after being stranded on 30 square kilometers for 2.5 FUCKING YEARS. they finally sent some consulate agent, coz it’s the French presidential elections. so i’m gonna finish this off first and only then i’ll be on iTermocil. thanks 😄

**Comment by godbout** [comment]
oh also @TomAnthony. not sure how you want to handle the repo now. are you completely out or you still want to have a say?

because what i usually do in my own projects is support only the two latest official language versions (usually the last one being the stable, the previous one being the security fixes and bugs updates). 

for iTermocil what i would do is release a v1 for Python 3 (probably 3.9) that would break for Python 2.7 most probably. like if people want iTermocil for 2.7, i guess we would need to do a versioned release in Homebrew. 

again, no knowledge of Python or the Python ecosystem here. i just know `pip` coz it sounds very funny in French.

**Comment by ruudk** [comment]
Let's keep it simple. When people want to run this on Python V2, they're on their own and can use an old tag.

The project should just be bumped to Python V3 as that's the default Python in Brew now.

**Comment by godbout** [comment]
but iTermocil is important! can't we add this to homebrew-core? i've added stuff homebrew-cask but never to core or taps. not sure about their policies. thought brew taps where kinda obsolete now?

**Comment by godbout** [comment]
policies here: https://docs.brew.sh/Acceptable-Formulae

**Comment by ruudk** [comment]
Let's first focus on fixing this, otherwise it will def never be accepted :)

**Comment by godbout** [comment]
ok i've made a PR with all the changes for Python3. what's needed now is to make a release. what i'm not sure about yet is how we want to install iTermocil. i wanna add it to homebrew-core and just do a `brew install itermocil`, but not sure yet whether it would be accepted or not. 

if not we may need to update @TomAnthony's brews/taps. i have no knowledge of personal taps, but it seems that one is currently outdated and would need some rework: https://github.com/TomAnthony/homebrew-brews/issues

**Comment by godbout** [comment]
> Let's first focus on fixing this, otherwise it will def never be accepted :)

i've updated to Python3. code works now. at least on my machine. not sure about M1. maybe best is that people install and try and report before we do a `homebrew` release.

**Comment by godbout** [comment]
@TomAnthony your brews' syntax need to be updated, currently they're not working. we can help here by:

1. updating your current tap. if it's ok with you, can you invite us as collaborators?
2. create a new tap, and then update the README with the new installation procedure. but then that kind of defeat the purpose of working on this current repo.

thanks.

**Comment by ruudk** [comment]
@godbout Didn't had my notifications set up so therefore I missed the PR. Just enabled them.
I reviewed it just yet and it seems to be good. 

@TomAnthony Please invite @godbout and @ruudk to the https://github.com/TomAnthony/homebrew-brews repo so that we can fix it there too.

**Comment by TomAnthony** [comment]
> ok. @TomAnthony any reason why this has to be installed through brew install TomAnthony/brews/itermocil and not directly in Homebrew core? brew install itermocil?

Just that it was a long time ago, and I tried to submit it to core a few times but couldn't get it accepted at the time.

> @TomAnthony Please invite @godbout and @ruudk to the https://github.com/TomAnthony/homebrew-brews repo so that we can fix it there too.

Done!

> oh also @TomAnthony. not sure how you want to handle the repo now. are you completely out or you still want to have a say?
> because what i usually do in my own projects is support only the two latest official language versions (usually the last one being the stable, the previous one being the security fixes and bugs updates).

Whilst I would love to stay involved, realistically I am too time poor and I don't want to hold things back. I'm happy for you to make decisions. :) I might want to be involved more again in the future, but in that case I would expect to be more of an equal collaborator rather then benevolent dictator. :)

Thanks for your work on this!

**Comment by godbout** [comment]
> Just that it was a long time ago, and I tried to submit it to core a few times but couldn't get it accepted at the time.

you're definitely smarter than me. it took three days to understand the doc and it seems you made the right choice. doesn't go to core.

> Done!

thank you!

> Whilst I would love to stay involved, realistically I am too time poor and I don't want to hold things back. I'm happy for you to make decisions. :) I might want to be involved more again in the future, but in that case I would expect to be more of an equal collaborator rather then benevolent dictator. :)

tbh i'm like you, barely have time (other priorities). but on any project i work the first i do is... run iTermocil 😄️ so we'll get it working at least. definitely no issue (probably would actually prefer) if you'd stay the benevolent dictator.

> Thanks for your work on this!

no worries. thanks for the amazing tool you created!

**Comment by godbout** [comment]
> Let's keep it simple. When people want to run this on Python V2, they're on their own and can use an old tag.
> 
> The project should just be bumped to Python V3 as that's the default Python in Brew now.

well, computers are not that simple 😅️

i've updated the brew and iTermocil's code. now here's the thing. by default there's not Python interpreter on macOS 12.3 anymore. but once you install Xcode and the component tools, Python is installed. but it's v3.8. installing through Homebrew with `brew install python` you can 3.9. but 3.10 is actually the latest version (https://www.python.org/downloads/), and you can install it with `brew install python@3.10`. 

so. i have no idea if the changes i've made would work on the 3 versions. plus afaik (which means nothing), we need to point to an interpreter in iTermocil's code. currently it points to `/usr/bin/python` which is the one installed with Xcode/tools. 

so bis. which one we go for? we'll have to edit the README i assume because afaik (again), iTermocil's users will **have** to install Python first. i'd say easiest would be just a `brew install python`, hence v3.9 (for now? maybe change in the future?). any thoughts?

**Comment by ruudk** [comment]
I'll have a look later. There are a lot of formulas that depend on Python 3 and then use a homebrew helper to get that specific path.

**Comment by godbout** [comment]
ok done. i can personally install through the usual instructions from the README (added the `brew install python` step tho). installs fine, then using it works well too. iTermocil itself should work with any Python3 version, but the setup is currently kinda hardcoded to 3.9. not sure how to handle that yet. will research if it's an issue for someone.

you guys please try and report! thank you.

**Comment by ruudk** [comment]
The problem has been fixed, please run `brew update && brew install tomanthony/brews/itermocil` to retry. 
If you encounter issues, please create a new issue.

**Comment by ruudk** [comment]
@godbout @DougTGP https://Ko-fi.com/home/coffeeshop?txid=f8f4359a-cbed-4a09-acb6-18facd883573&mode=public&img=ogiboughtsomeone as promised, cheers 🍻

**Comment by godbout** [comment]
hey @ruudk thanks but you did some great work too. so back to you. where's your own ko-fi page? :D
